### PR TITLE
Mlops development

### PR DESCRIPTION
**Pull request by mizaqq**
PR contains:
- Hydra configs for whole project, including dynamic model initialization, training parameters etc.
- PyTorch Lightning module for easier training setup
- Many changes requested in previous PR

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -4,6 +4,7 @@ version = "0.1.0"
 description = ""
 authors = ["Michał Zaręba <michal.zareba@softwaremill.pl>"]
 readme = "README.md"
+package-mode = false
```
Comment (adamwawrzynski) [inline pyproject.toml]: What's the rationale behind this line? Why you don't want to install package itself?

File: src/model/train.py [hunk 2]
```diff
@@ -1,19 +1,29 @@
...
 
 class Model:
-    def __init__(self, model: torch.nn.Module, device: torch.device, data_loader: Dataloader) -> None:
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        data_loader: Dataloader,
+        optimizer: torch,
+        loss_fn: torch = torch.nn.CrossEntropyLoss(),
+        device: torch.device = None,
```
Comment (adamwawrzynski) [inline src/model/train.py]: `mypy` should give you an error in this line.
```suggestion
        device: torch.device | None = None,
```

File: src/model/train.py [hunk 3]
```diff
@@ -1,19 +1,29 @@
...
 
 
 class Model:
-    def __init__(self, model: torch.nn.Module, device: torch.device, data_loader: Dataloader) -> None:
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        data_loader: Dataloader,
+        optimizer: torch,
+        loss_fn: torch = torch.nn.CrossEntropyLoss(),
```
Comment (adamwawrzynski) [inline src/model/train.py]: Those two types `torch` are incorrect. You should provide base classes for optimizer and loss. And `mypy` should give you an error in those places.

File: src/model/lightning.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        if 'accuracy' in self.metrics:
+            self.log('Accuracy', accuracy_score(labels, preds))
+        if 'precision' in self.metrics:
+            self.log('Precision', precision_score(labels, preds, average="weighted"))
+        if 'recall' in self.metrics:
+            self.log('Recall', recall_score(labels, preds, average="weighted"))
+        if 'f1' in self.metrics:
+            self.log('F1', f1_score(labels, preds, average="weighted"))
+        if 'classification_report' in self.metrics:
+            self.log('classification_report', classification_report(labels, preds))
```
Comment (adamwawrzynski) [inline src/model/lightning.py]: You could extract this method as a staticmethod or normal function to return dict with metric name and metric values or just metric value. And in this method you could call it and log result to the training. The point is to separate logic of returning correct value from the logging process. It would be much easier to test this function without mocking the `.log` method.

File: src/model/train.py [hunk 5]
```diff
@@ -22,25 +32,59 @@ def train_model(self, lr: float = 0.001, momentum: float = 0.9, epochs: int = 10
...
+            print(f'Recall: {recall}')
+            mlflow.log_metric("Recall", recall)
+        if 'f1' in metrics:
+            f1 = f1_score(labels_list, predicted_labels, average="weighted")
+            print(f'F1 Score: {f1}')
+            mlflow.log_metric("F1 Score", f1)
+        if 'classification_report' in metrics:
+            c_report = classification_report(labels_list, predicted_labels, average="weighted")
+            print(f'Classification Report: {c_report}')
+            mlflow.log_text("Classification Report", c_report)
```
Comment (adamwawrzynski) [inline src/model/train.py]: Same as for lightning module. You could extract this logic into a staticmethod or a function and return the proper value.

File: src/utils/mlflow.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import os
+
+import mlflow
+
+
+class MLFlowHandler:
```
Comment (adamwawrzynski) [inline src/utils/mlflow.py]: This name is misleading. This class only starts and stops MLFlow runs. You are logging all artifacts and metadata directly using `mlflow` which is confusing. My suggestions:
1. Rename class to something like `MLFlowRunManager`
2. Add method to return mlflow object on which user can operate after starting the experiment.
3. Reimplement methods from `mlflow` (e.g. `log_dict`) and use decorator to automatically start and stop mlflow run tracking (if you want to add something to the mlflow run that was closed you have to provide the ID of the experiment).

Version 3 adds a lot of new code to write but might be good choice if you have custom logging functions or logic. Just as a reference how the point 3. might be implemented. This is decorator method of a class.
```python
def log_to_mlflow(func: Callable[Concatenate[Any, P], T]) -> Method[P, T]:  # type: ignore
        @wraps(func)
        def wrapper(self: MLFlowLogger, *args: P.args, **kwargs: P.kwargs) -> T:
            with mlflow.start_run(
                run_id=self.run_id,
                tags=self.tags,
                description=self.description,
                run_name=self.run_name,
            ):
                return func(self, *args, **kwargs)

        return wrapper  # type: ignore
```

**Review by adamwawrzynski — APPROVED** [review]
Overall LGTM

File: pyproject.toml (hunk 1 above)
Comment (mizaqq) [reply pyproject.toml]: My idea for using poetry is just dependency management, not publishing the code as package itself, second thing was that without this line I get error `Warning: The current project could not be installed: No file/folder found for package plant-disease`
This is due to naming main folder as src, changed to including package 'src' now .

File: src/model/lightning.py (hunk 4 above)
Comment (mizaqq) [reply src/model/lightning.py]: We should talk about this, this function is derived from Lightning module and its called when training model. I couldnt find option to get metric values as dict utilizing the module, not just calling the calculated model and generating result bypassing Lightning

File: src/utils/mlflow.py (hunk 6 above)
Comment (mizaqq) [reply src/utils/mlflow.py]: [971a714](https://github.com/mizaqq/plant-disease/pull/2/commits/971a71453fd5d347f3499ff5474ce1c04ff4fed9)
Renamed the class, now the class is used to log to mlflow

File: src/model/train.py (hunk 5 above)
Comment (mizaqq) [reply src/model/train.py]: [971a714](https://github.com/mizaqq/plant-disease/pull/2/commits/971a71453fd5d347f3499ff5474ce1c04ff4fed9)
changed

File: src/model/train.py (hunk 3 above)
Comment (mizaqq) [reply src/model/train.py]: [fixed](https://github.com/mizaqq/plant-disease/pull/2/commits/971a71453fd5d347f3499ff5474ce1c04ff4fed9)
